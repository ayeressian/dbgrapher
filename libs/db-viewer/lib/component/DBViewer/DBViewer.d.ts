type DbViewerEventMap = import("./event").DbViewerEventMap;
type Schema = import("../../schema").Schema;
type Viewport = import("../../store/view").Viewport;

export default class DbViewer extends HTMLElement {
  setSchema(schema: Schema, viewport: Viewport): void;
  getSchema(): Schema;

  addEventListener<K extends keyof DbViewerEventMap>(
    type: K,
    listener: (this: DbViewer, ev: DbViewerEventMap[K]) => unknown,
    options?: boolean | AddEventListenerOptions
  ): void;

  removeEventListener<K extends keyof DbViewerEventMap>(
    type: K,
    listener: (this: HTMLFormElement, ev: DbViewerEventMap[K]) => unknown,
    options?: boolean | EventListenerOptions
  ): void;
}

export * from "../../schema";
export * from "./event";
export { default as Point } from "../../point";
