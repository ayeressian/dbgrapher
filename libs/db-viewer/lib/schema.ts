import type Point from "./point";

export interface FkSchema {
  table: string;
  column: string;
}
export interface ColumnNoneFkSchema {
  name: string;
  type: string;
  pk?: boolean;
  uq?: boolean;
  nn?: boolean;
}

export declare interface ColumnFkSchema
  extends Omit<ColumnNoneFkSchema, "type"> {
  fk: FkSchema;
}

export type ColumnSchema = ColumnFkSchema | ColumnNoneFkSchema;

export interface TableSchema {
  name: string;
  pos: Point;
  columns: ColumnSchema[];
}

export interface AnnotationSchema {
  pos: Point;
  title?: string;
  text?: string;
  color?: string;
  width: number;
  height: number;
}

export interface Schema {
  tables: TableSchema[];
  annotations?: AnnotationSchema[];
  arrangement?: TableArrang;
  viewport?: Viewport;
  viewWidth?: number;
  viewHeight?: number;
}

export const enum Viewport {
  noChange = "noChange",
  centerByTablesWeight = "centerByTablesWeight",
  center = "center",
  centerByTables = "centerByTables",
}

export enum TableArrang {
  spiral = "spiral",
  default = "default",
}

export const isColumnFk = (column: ColumnSchema): column is ColumnFkSchema => {
  return (column as ColumnFkSchema).fk !== undefined;
};
