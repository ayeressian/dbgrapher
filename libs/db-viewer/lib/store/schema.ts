import { get, writable, type Readable } from "svelte/store";
import type { Schema } from "../schema";
import type { Store } from "./store";
import type { Viewport } from "./view";

export type RectSize = {
  width: number;
  height: number;
};

export class SchemaStore {
  private store: Store;
  schema = writable<Schema>({ tables: [] });
  // allTableSizesSetStore: Readable<boolean>;

  constructor(store: Store) {
    this.store = store;
  }

  getSchema(): Schema {
    const schema = get(this.schema);
    this.store.table.updateTablePoses(schema.tables);
    return schema;
  }

  setSchema(schema: Schema, viewport: Viewport = "center") {
    schema = JSON.parse(JSON.stringify(schema));
    this.store?.view.viewportStore.set(viewport);
    const tableStore = this.store.table;
    tableStore.tableSizes.clear();
    tableStore.tablePoses.clear();
    tableStore.tableToHighlight.clear();
    tableStore.tableFromHighlight.clear();
    schema.tables.forEach((table) => {
      const tableSizeWritable = writable<RectSize>();
      tableStore.tableSizes.set(table.name, tableSizeWritable);
      const tablePosWritable = writable(table.pos);
      tableStore.tablePoses.set(table.name, tablePosWritable);
      tablePosWritable.subscribe(this.store!.view.getUpdateViewBound());
      tableStore.tableToHighlight.set(table.name, writable(""));
      tableStore.tableFromHighlight.set(table.name, writable(""));
    });
    this.schema.set({tables: []});
    setTimeout(() => this.schema.set(schema));    

    // allTableSizesSetStore = derived(
    //   [...tableSizes.values()],
    //   (tableSizes) => {
    //     return tableSizes.every((tableSize) => {
    //       return tableSize.height > 0 && tableSize.width > 0;
    //     });
    //   }
    // );
  }
}
