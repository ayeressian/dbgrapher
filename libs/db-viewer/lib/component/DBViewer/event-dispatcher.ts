import type Point from "lib/point";
import type { RectSize } from "lib/store/schema";
import {
  RelationClickEvent,
  RelationContextMenuEvent,
  RelationDblClickEvent,
  TableClickEvent,
  TableContextMenuEvent,
  TableDblClickEvent,
  ViewportClickEvent,
  ZoomEvent,
  type RelationData,
  type TableData,
  type ZoomData,
} from "./event";

export default class EventDispatcher {
  constructor(private dbViewer: DbViewer) {}
  onTableClick = (tableData: TableData) => {
    this.dbViewer.dispatchEvent(new TableClickEvent(tableData));
  };

  onTableDblClick = (tableData: TableData) => {
    this.dbViewer.dispatchEvent(new TableDblClickEvent(tableData));
  };

  onTableContextMenu = (tableData: TableData) => {
    this.dbViewer.dispatchEvent(new TableContextMenuEvent(tableData));
  };

  onRelationClick = (realtionData: RelationData) => {
    this.dbViewer.dispatchEvent(new RelationClickEvent(realtionData));
  };

  onRelationDblClick = (realtionData: RelationData) => {
    this.dbViewer.dispatchEvent(new RelationDblClickEvent(realtionData));
  };

  onRelationContextMenu = (realtionData: RelationData) => {
    this.dbViewer.dispatchEvent(new RelationContextMenuEvent(realtionData));
  };

  onViewportClick = (point: Point) => {
    this.dbViewer.dispatchEvent(new ViewportClickEvent(point));
  };

  onZoom = (zoomData: ZoomData) => {
    this.dbViewer.dispatchEvent(new ZoomEvent(zoomData));
  };
}
