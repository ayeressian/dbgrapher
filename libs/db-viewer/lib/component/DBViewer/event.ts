import type { RectSize } from "../../store/schema";
import type Point from "../../point";

export type TableData = {
  name: string;
  pos: Point;
  size: RectSize;
};

export type RelationData = {
  fromTable: string;
  toTable: string;
  fromColumn: string;
  toColumn: string;
};

export type ZoomData = {
  newZoom: number;
  oldZoom: number;
  target: Point;
};

export default TableData;

export class TableClickEvent extends CustomEvent<TableData> {
  constructor(tableData: TableData) {
    super("tableClick", { detail: tableData });
  }
}

export class TableDblClickEvent extends CustomEvent<TableData> {
  constructor(tableData: TableData) {
    super("tableDblClick", { detail: tableData });
  }
}

export class TableContextMenuEvent extends CustomEvent<TableData> {
  constructor(tableData: TableData) {
    super("tableContextMenu", { detail: tableData });
  }
}

export class RelationClickEvent extends CustomEvent<RelationData> {
  constructor(relationData: RelationData) {
    super("relationClick", { detail: relationData });
  }
}

export class RelationDblClickEvent extends CustomEvent<RelationData> {
  constructor(relationData: RelationData) {
    super("relationDblClick", { detail: relationData });
  }
}

export class RelationContextMenuEvent extends CustomEvent<RelationData> {
  constructor(relationData: RelationData) {
    super("relationContextMenu", { detail: relationData });
  }
}

export class ViewportClickEvent extends CustomEvent<Point> {
  constructor(point: Point) {
    super("viewportClick", { detail: point });
  }
}

export class TableMoveEvent extends CustomEvent<TableData> {
  constructor(tableData: TableData) {
    super("tableMove", { detail: tableData });
  }
}

export class TableMoveEndEvent extends CustomEvent<TableData> {
  constructor(tableData: TableData) {
    super("tableMoveEnd", { detail: tableData });
  }
}

export class ZoomEvent extends CustomEvent<ZoomData> {
  constructor(zoomData: ZoomData) {
    super("zoom", {
      detail: zoomData,
    });
  }
}

export interface DbViewerEventMap extends HTMLElementEventMap {
  tableClick: TableClickEvent;
  tableDblClick: TableDblClickEvent;
  tableContextMenu: TableContextMenuEvent;
  relationClick: RelationClickEvent;
  relationDblClick: RelationDblClickEvent;
  relationContextMenu: RelationContextMenuEvent;
  viewportClick: ViewportClickEvent;
  tableMove: TableMoveEvent;
  tableMoveEnd: TableMoveEndEvent;
  zoom: ZoomEvent;
}
