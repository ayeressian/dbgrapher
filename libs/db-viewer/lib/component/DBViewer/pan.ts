import type Point from "lib/point";
import { get, type Writable } from "svelte/store";

class Pan {
  private lastPos?: Point;

  private prevMouseCordX!: number;
  private prevMouseCordY!: number;

  constructor(
    private viewPosStore: Writable<Point>,
    private zoomStore: Writable<number>,
    private onPanStart?: () => void,
    private onPanEnd?: () => void
  ) {}

  private noneTableAndMinmapEvent(event: Event): boolean {
    return !event.composedPath().some((item) => {
      const htmlElement = item as HTMLElement;
      return (
        htmlElement.classList?.contains("minimap") ||
        htmlElement.tagName === "TABLE"
      );
    });
  }

  private onMouseMove = (event: MouseEvent) => {
    event.preventDefault();
    if (this.noneTableAndMinmapEvent(event)) {
      const deltaX = event.clientX - this.prevMouseCordX;
      const deltaY = event.clientY - this.prevMouseCordY;
      this.prevMouseCordY = event.clientY;
      this.prevMouseCordX = event.clientX;
      const zoom = get(this.zoomStore);
      this.viewPosStore.update((point: Point) => {
        point.x -= deltaX * zoom;
        point.y -= deltaY * zoom;
        return point;
      });
    }
  };

  onMousedown = (event: MouseEvent) => {
    if (event.button === 0 && this.noneTableAndMinmapEvent(event)) {
      this.onPanStart && this.onPanStart();
      this.prevMouseCordX = event.clientX;
      this.prevMouseCordY = event.clientY;
      (event.target as SVGSVGElement).addEventListener(
        "mousemove",
        this.onMouseMove
      );
    }
  };

  onMouseup = (event: MouseEvent) => {
    this.onPanEnd && this.onPanEnd();
    (event.target as SVGSVGElement).removeEventListener(
      "mousemove",
      this.onMouseMove
    );
  };
}

export default Pan;
