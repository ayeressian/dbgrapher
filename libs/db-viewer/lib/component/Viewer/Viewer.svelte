<script lang="ts">
  import type { Schema } from "../../schema";
  import { getContext } from "svelte";
  import Table from "../Table/Table.svelte";
  import Relation from "../Relation/Relation.svelte";
  import type { Viewport } from "../../store/view";
  import type Point from "../../point";
  import Minimap from "../Minimap/Minimap.svelte";
  import type { Store } from "../../store/store";
  import ZoomButtons from "../ZoomButtons/ZoomButtons.svelte";
  import Pan from "./pan";
  import type EventDispatcher from "../DBViewer/event-dispatcher";
  import type TableData from "../DBViewer/event";
  import Zoom from "../../zoom";

  export let eventDispatcher: EventDispatcher;

  const store = getContext<Store>("store");

  const schemaStore = store.schema.schema;

  export const getSchema = (): Schema => store.schema.getSchema();
  export const setSchema = (schema: Schema, viewport: Viewport) => {
    store.schema.setSchema(schema, viewport);
  };

  const SCROLL_TO_ZOOM_MULTIPLIER = 0.01;

  let svgElem: SVGSVGElement;

  const zoomStore = store.view.zoom;
  const viewSizeStore = store.view.viewSize;
  const viewPosStore = store.view.viewPos;
  const viewSizeAfterZoomStore = store.view.viewSizeAfterZoom;
  const relationsStore = store.relation.relations;

  function getSvgMousePos(event: MouseEvent) {
    const pt = svgElem.createSVGPoint();
    pt.x = event.clientX;
    pt.y = event.clientY;
    const svgPoint = pt.matrixTransform(svgElem?.getScreenCTM()?.inverse());
    return { x: svgPoint.x, y: svgPoint.y };
  }

  function zoomHandler(event: WheelEvent) {
    event.preventDefault();

    const bb = svgElem.getBoundingClientRect();
    const mousePos: Point = {
      x: event.clientX - bb.left,
      y: event.clientY - bb.top,
    };
    const zoomAmount = event.deltaY * SCROLL_TO_ZOOM_MULTIPLIER;
    const oldZoom = $zoomStore;
    console.log("AAAA", event.clientX, mousePos);
    const result = new Zoom(
      zoomAmount,
      mousePos,
      oldZoom,
      $viewSizeStore,
      $viewPosStore
    ).call();

    viewPosStore.set(result.pos);

    zoomStore.set(result.zoom);
    eventDispatcher.onZoom({
      newZoom: result.zoom,
      oldZoom,
      target: getSvgMousePos(event),
    });
  }

  function scrollOperation(scrollXAmount: number, scrollYAmount: number) {
    viewPosStore.update((viewPosVal) => {
      viewPosVal.x += scrollXAmount;
      viewPosVal.y += scrollYAmount;
      return viewPosVal;
    });
  }

  function scrollHandler(event: WheelEvent) {
    event.preventDefault();
    scrollOperation(event.deltaX, event.deltaY);
  }

  function mousewheel(event: WheelEvent) {
    if (event.ctrlKey) {
      zoomHandler(event);
      return;
    }
    scrollHandler(event);
  }

  function onTableMouseDown({ name }: TableData) {
    const index = $schemaStore.tables.findIndex((table) => table.name === name);
    const [table] = $schemaStore.tables.splice(index, 1);
    $schemaStore.tables = [...$schemaStore.tables, table];
  }

  function click(event: MouseEvent) {
    eventDispatcher.onViewportClick(getSvgMousePos(event));
  }

  let paning = false;

  function onPanStart() {
    paning = true;
  }
  function onPanEnd() {
    paning = false;
  }
  const drag = new Pan(viewPosStore, zoomStore, onPanStart, onPanEnd);
</script>

<!-- svelte-ignore a11y-click-events-have-key-events -->
<svg
  class="view"
  bind:this={svgElem}
  on:wheel={mousewheel}
  on:click={click}
  on:mousedown={drag.onMousedown}
  on:mouseup={drag.onMouseup}
  class:pan={paning}
  version="2"
  xmlns="http://www.w3.org/2000/svg"
  viewBox={`${$viewPosStore.x} ${$viewPosStore.y} ${$viewSizeAfterZoomStore.width} ${$viewSizeAfterZoomStore.height}`}
>
  {#each $relationsStore ?? [] as relation (relation.key)}
    <Relation
      {...relation}
      click={eventDispatcher.onRelationClick}
      dblClick={eventDispatcher.onRelationDblClick}
      contextMenu={eventDispatcher.onRelationContextMenu}
    />
  {/each}
  {#each $schemaStore?.tables ?? [] as { name } (name)}
    <Table
      {name}
      mouseDown={onTableMouseDown}
      click={eventDispatcher.onTableClick}
      dblClick={eventDispatcher.onTableDblClick}
      contextMenu={eventDispatcher.onTableContextMenu}
    />
  {/each}
</svg>
<Minimap />
<ZoomButtons />
