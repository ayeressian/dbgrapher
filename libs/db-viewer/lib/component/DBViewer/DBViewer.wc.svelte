<svelte:options customElement='db-viewer'/>

<script lang="ts">
  import type { Schema } from "../../schema";
  import { onMount, setContext } from "svelte";
  import type { Viewport } from "../../store/view";
  import { Store } from "../../store/store";
  import EventDispatcher from "./event-dispatcher";
  import type DbViewer from "./DBViewer";
  import Viewer from "../Viewer/Viewer.svelte";

  const store = new Store();
  setContext("store", store);

  export const getSchema = (): Schema => {
    const copy = JSON.parse(JSON.stringify(store.schema.getSchema()));
    return copy;
  }
  export const setSchema = (schema: Schema, viewport: Viewport) => {
    store.schema.setSchema(schema, viewport);
  };

  let container: HTMLDivElement;
  let eventDispatcher: EventDispatcher;

  onMount(() => {
    const shadowRoot = container.parentNode as ShadowRoot;

    const host = shadowRoot.host as DbViewer;

    eventDispatcher = new EventDispatcher(host);
    store.view.viewSize.set({
      width: host.clientWidth,
      height: host.clientHeight,
    });

    const svgResizeObserver = new ResizeObserver((entries) => {
      store.view.viewSize.set({
        width: host.clientWidth,
        height: host.clientHeight,
      });
    });
    svgResizeObserver.observe(host);
  });
</script>

<style>
  :host {
    display: block;
  }
  #container {
    width: 100%;
    height: 100%;
    position: relative;
  }

  :global(*) {
    font-family: var(--font-family, Arial, sans-serif);
    color: var(--color, #333);
    --table-boarder-color: #a4a2a3;
    --table-color: #a4a2a3;
    --annotation-color: #ffffff;
  }

  :global(svg) {
    background-color: #E3E3E3;
  }
</style>

<div id="container" bind:this={container}>
  <Viewer {eventDispatcher} />
</div>
