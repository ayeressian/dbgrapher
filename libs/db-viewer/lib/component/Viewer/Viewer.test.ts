import { fireEvent, render, type RenderResult } from "@testing-library/svelte";
import Viewer from "./Viewer.svelte";
import school from "../../../src/school";
import { beforeEach, describe, expect, it, SpyInstance, vi } from "vitest";
import { Store } from "../../store/store";
import * as zoomImport from "../../zoom";
import { Point } from "../DBViewer/DBViewer";
import { RectSize } from "lib/store/schema";

function getStore() {
  const store = new Store();
  store.schema.setSchema(school);
  store.table.getTableSize("school")?.set({
    width: 200,
    height: 300,
  });
  store.table.getTableSize("student")?.set({
    width: 200,
    height: 400,
  });
  store.view.viewSize.set({
    width: 800,
    height: 1000,
  });
  return store;
}

describe(Viewer.name, () => {
  let component: RenderResult<Viewer>;

  const viewWidth = 800,
    viewHeight = 1000;

  global.ResizeObserver = vi.fn().mockImplementation((func) => {
    return {
      observe: () => {
        func([
          { borderBoxSize: [{ inlineSize: viewWidth, blockSize: viewHeight }] },
        ]);
      },
    };
  });

  Object.defineProperty(global.SVGSVGElement.prototype, "createSVGPoint", {
    writable: true,
    value: vi.fn().mockImplementation(() => ({
      x: 0,
      y: 0,
      matrixTransform: vi.fn().mockImplementation(() => ({
        x: 0,
        y: 0,
      })),
    })),
  });

  beforeEach(async () => {
    const context = new Map();
    context.set("store", getStore());

    component = render(Viewer, {
      props: {
        eventDispatcher: {
          onZoom: vi.fn(),
        },
      },
      context,
    });
  });

  it("should render properly", () => {
    expect(component).toMatchSnapshot();
  });

  describe("scroll view", () => {
    let svgElem: SVGElement;
    const deltaX = 100,
      deltaY = 200;
    beforeEach(() => {
      svgElem = document.querySelector("svg") as SVGElement;
      fireEvent.wheel(svgElem, { deltaX, deltaY });
    });
    it("viewport should change", () => {
      expect(svgElem.getAttribute("viewBox")).toEqual(
        `${deltaX} ${deltaY} ${viewWidth} ${viewHeight}`
      );
    });
  });

  describe("zoom view", () => {
    let svgElem: SVGElement;
    const clientX = 300,
      clientY = 400,
      deltaY = -2;

    let ZoomClass: SpyInstance<
      [
        amount: number,
        target: Point,
        oldZoom: number,
        viewSize: RectSize,
        viewPos: Point
      ],
      zoomImport.default
    >;

    beforeEach(async () => {
      Object.defineProperty(
        global.SVGGraphicsElement.prototype,
        "getScreenCTM",
        {
          writable: true,
          value: vi.fn().mockReturnValue({
            inverse: vi.fn(),
          }),
        }
      );

      Object.defineProperty(global.SVGSVGElement.prototype, "createSVGPoint", {
        writable: true,
        value: vi.fn().mockImplementation(() => ({
          x: clientX,
          y: clientY,
          matrixTransform: vi.fn().mockImplementation(() => ({
            x: clientX,
            y: clientY,
          })),
        })),
      });
      const OriginalZoom = zoomImport.default;

      ZoomClass = vi.spyOn(zoomImport, "default");
      ZoomClass.mockImplementation(
        (
          amount: number,
          target: Point,
          oldZoom: number,
          viewSize: RectSize,
          viewPos: Point
        ) => {
          return new OriginalZoom(amount, target, oldZoom, viewSize, viewPos);
        }
      );

      svgElem = document.querySelector("svg") as SVGElement;
      await fireEvent.wheel(svgElem, {
        clientX,
        clientY,
        deltaY,
        ctrlKey: true,
      });
    });

    it("zoom should be called", async () => {
      expect(ZoomClass).toHaveBeenCalledWith(
        -0.02,
        { x: clientX, y: clientY },
        1,
        { width: 800, height: 1000 },
        { x: 100, y: 200 }
      );
    });
  });
});
