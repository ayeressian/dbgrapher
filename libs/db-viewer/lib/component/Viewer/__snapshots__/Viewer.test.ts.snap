// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Viewer > should render properly 1`] = `
{
  "component": Viewer {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {
        "store" => Store {
          "relation": RelationStore {
            "relations": {
              "subscribe": [Function],
            },
            "store": [Circular],
          },
          "schema": SchemaStore {
            "schema": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "store": [Circular],
          },
          "table": TableStore {
            "tableFromHighlight": Map {
              "school" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "class" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "student" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
            },
            "tablePoses": Map {
              "school" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "class" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "student" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
            },
            "tableSizes": Map {
              "school" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "class" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "student" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
            },
            "tableToHighlight": Map {
              "school" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "class" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
              "student" => {
                "set": [Function],
                "subscribe": [Function],
                "update": [Function],
              },
            },
          },
          "view": ViewStore {
            "store": [Circular],
            "viewBound": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "viewBoundSize": {
              "subscribe": [Function],
            },
            "viewPos": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "viewSize": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "viewSizeAfterZoom": {
              "subscribe": [Function],
            },
            "viewportApply": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "viewportStore": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
            "zoom": {
              "set": [Function],
              "subscribe": [Function],
              "update": [Function],
            },
          },
        },
      },
      "ctx": [
        {
          "onZoom": [MockFunction spy],
        },
        <svg
          class="view svelte-1s0sr4k"
          version="2"
          viewBox="0 0 800 1000"
          xmlns="http://www.w3.org/2000/svg"
        >
          
        </svg>,
        false,
        {
          "tables": [],
        },
        {
          "x": 0,
          "y": 0,
        },
        {
          "height": 1000,
          "width": 800,
        },
        [],
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "subscribe": [Function],
        },
        {
          "subscribe": [Function],
        },
        [Function],
        [Function],
        [Function],
        Pan {
          "onMouseMove": [Function],
          "onMousedown": [Function],
          "onMouseup": [Function],
          "onPanEnd": [Function],
          "onPanStart": [Function],
          "prevMouseCordX": undefined,
          "prevMouseCordY": undefined,
          "viewPosStore": {
            "set": [Function],
            "subscribe": [Function],
            "update": [Function],
          },
          "zoomStore": {
            "set": [Function],
            "subscribe": [Function],
            "update": [Function],
          },
        },
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "eventDispatcher": 0,
      },
      "root": <div>
        <svg
          class="view svelte-1s0sr4k"
          version="2"
          viewBox="0 0 800 1000"
          xmlns="http://www.w3.org/2000/svg"
        >
          
        </svg>
         
        <svg
          class="minimap svelte-1xf01od"
          height="200"
          version="2"
          viewBox="0 0 822 1000"
          width="164.4"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            class="view-minimap svelte-1xf01od"
            height="1000"
            width="800"
            x="0"
            y="0"
          />
        </svg>
         
        <button
          class="zoom-btn in pure-button svelte-fpylet"
        >
          +
        </button>
         
        <button
          class="zoom-btn out pure-button svelte-fpylet"
        >
          -
        </button>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$events_def": undefined,
    "$$prop_def": undefined,
    "$$set": [Function],
    "$$slot_def": undefined,
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <svg
        class="view svelte-1s0sr4k"
        version="2"
        viewBox="0 0 800 1000"
        xmlns="http://www.w3.org/2000/svg"
      >
        
      </svg>
       
      <svg
        class="minimap svelte-1xf01od"
        height="200"
        version="2"
        viewBox="0 0 822 1000"
        width="164.4"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          class="view-minimap svelte-1xf01od"
          height="1000"
          width="800"
          x="0"
          y="0"
        />
      </svg>
       
      <button
        class="zoom-btn in pure-button svelte-fpylet"
      >
        +
      </button>
       
      <button
        class="zoom-btn out pure-button svelte-fpylet"
      >
        -
      </button>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
